### PR TITLE
Refactor config, add `settings` command

### LINKED ISSUE #158: [Bug] Running `settings help` returns nothing
**Fill out general information**
OS: Linux
BeamMP-Server Version: 3.1.1

**Describe the bug**
Running the `settings help` command, as suggested within the `help` command, returns nothing

**To Reproduce**
1. Run `settings help`
2. Observe nothing happen

**Expected behavior**
Something to happen

### PR DESCRIPTION
**Pull request by jimkoen**
Fix #158

### CODE REVIEW

**Comment by jimkoen** [comment]
todo:
- support concurrent access to config
- cleanup in multiple places
- change settings access to new settings.get... everywhere

**Comment by lionkor** [comment]
@jimkoen what's the status on this?

**Comment by jimkoen** [comment]
> @jimkoen what's the status on this?

I have a working version on my end where I need to fix 1-2 remaining build errors related to formatting of console output.

However as discussed privately I'm held up by exams, so this  won't land before the week of the 29th of April.
If I'm holding something up, please tell me, I might be able to reschedule, but otherwise it would be very considerate if you could respect my exam schedule.

**Comment by lionkor** [comment]
Thank you for the info

**Comment by jimkoen** [comment]
@lionkor Should be ready for review.

Unfortunately I cannot reproduce the test failures on Ubuntu 20.04 on my end. Tests work fine on FreeBSD (and other OS's as indicated by the Actions result). For now I think it'd make sense to start the review, so the issue doesn't become any more stale than it already is.

File: include/TConfig.h
Comment (lionkor) [inline include/TConfig.h]: delete pls

File: include/TSettings.h
Comment (lionkor) [inline include/TSettings.h]: write perms for changing it

File: include/TSettings.h
Comment (lionkor) [inline include/TSettings.h]: read+write

File: include/TSettings.h
Comment (lionkor) [inline include/TSettings.h]:
```suggestion
        READ_ONLY, // Value can be read from console
        READ_WRITE, // Value can be read and written to from console
        NO_ACCESS, // Value is inaccessible from console (no read OR write)
```

File: include/TSettings.h
Comment (lionkor) [inline include/TSettings.h]:
```suggestion
```

File: include/TSettings.h
Comment (lionkor) [inline include/TSettings.h]:
```suggestion
    void set(Key key, const std::string& value) {
```

File: include/TSettings.h
Comment (lionkor) [inline include/TSettings.h]:
```suggestion
    const std::unordered_map<ComposedKey, SettingsAccessControl> GetAccessControlMap() const {
```

Comment (lionkor) [reply include/TSettings.h]: something like that pls

File: include/TSettings.h
Comment (lionkor) [inline include/TSettings.h]:
```suggestion
    void setConsoleInputAccessMapping(const ComposedKey& keyName, const std::string& value) {
```

File: include/TSettings.h
Comment (lionkor) [inline include/TSettings.h]:
```suggestion
```

File: src/TConsole.cpp
Comment (lionkor) [inline src/TConsole.cpp:373]: use `EnsureArgsCount` here for consistency if possible, not sure rn if it has an overload for this use case, but im pretty sure it has a min+max variant

File: include/Common.h
Comment (lionkor) [inline include/Common.h]: could we please delete this whole old TSettings now that its not used? :)

**Review by lionkor — CHANGES_REQUESTED** [review]
Looks great, thank you. I have yet to test this, but rn ive marked some places that need fixing up.

You can use the suggestions ive made to commit the changes directly, either one by one or as a batch.

File: include/TSettings.h
Comment (lionkor) [inline include/TSettings.h]: Please rename into `Settings.h` and move all non-template fields and methods into a `src/Settings.cpp` file

**Review by lionkor — CHANGES_REQUESTED** [review]

File: include/TSettings.h
Comment (jimkoen) [reply include/TSettings.h]: wdym, all of these should get write access from runtime?

**Comment by jimkoen** [comment]
I implemented the changes. Currently I'm looking into a way to test Ubuntu 20.04 locally, as gh actions has stopped working for this branch, despite the fact that I haven't made any changes to the actions scripts.

**Comment by lionkor** [comment]
run this in your branch:

```
docker run -it --rm -v .:/test ubuntu:20.04
```

and then just run all the scrips in `./scripts/ubuntu-20.04/`  (you can skip the 3-build script and only build the tests). I made this *as* simple as possible without having to write it twice.

File: src/TConfig.cpp
Comment (lionkor) [inline src/TConfig.cpp]: This todo either has to be done or the comment needs to be removed

File: src/TConsole.cpp
Comment (lionkor) [inline src/TConsole.cpp]: let's use `{}::{}` here for consistency

File: include/Settings.h
Comment (lionkor) [inline include/Settings.h:48]: these can probably(?) be moved into the .cpp file

File: include/Settings.h
Comment (lionkor) [inline include/Settings.h]: move the `= ...`  half of this to the .cpp file to avoid full project recompilation every time we change a default setting.

File: include/Settings.h
Comment (lionkor) [inline include/Settings.h]: move the = ... half of this to the .cpp file as well, and possibly rename if there's maybe a more descriptive name

File: include/Settings.h
Comment (lionkor) [inline include/Settings.h]: these can also go in the .cpp

File: include/Settings.h
Comment (lionkor) [inline include/Settings.h:59]: this can be moved to the .cpp as well most likely? check please.

**Review by lionkor — CHANGES_REQUESTED** [review]

File: include/Settings.h
Comment (lionkor) [reply include/Settings.h:48]: nvm

File: include/Settings.h
Comment (lionkor) [reply include/Settings.h:59]: nvm

**Comment by jimkoen** [comment]
ready for review

File: src/TConfig.cpp
Comment (lionkor) [inline src/TConfig.cpp]: delete

File: src/TConsole.cpp
Comment (lionkor) [inline src/TConsole.cpp]:
```suggestion
        beammp_errorf("Unknown argument for command 'settings': {}", args.front());
```

File: src/TConsole.cpp
Comment (lionkor) [inline src/TConsole.cpp]: delete

File: src/TConsole.cpp
Comment (lionkor) [inline src/TConsole.cpp]: do or delete

**Review by lionkor — CHANGES_REQUESTED** [review]

**Comment by lionkor** [comment]
And please run clang-format

**Comment by lionkor** [comment]
A few things I noticed while testing:

1. I get this when I run `settings list`: image

2. When I enter an incomplete command I get a difficult to understand error: image

3. AuthKey is gettable and settable:  image

**Comment by lionkor** [comment]
Also, we need to do this:

```patch
diff --git a/src/THeartbeatThread.cpp b/src/THeartbeatThread.cpp
index a7f37c6..3e6699c 100644
--- a/src/THeartbeatThread.cpp
+++ b/src/THeartbeatThread.cpp
@@ -40,8 +40,8 @@ void THeartbeatThread::operator()() {
     static std::chrono::high_resolution_clock::time_point LastUpdateReminderTime = std::chrono::high_resolution_clock::now();
     bool isAuth = false;
     std::chrono::high_resolution_clock::duration UpdateReminderTimePassed;
-    auto UpdateReminderTimeout = ChronoWrapper::TimeFromStringWithLiteral(Application::Settings.getAsString(Settings::Key::Misc_UpdateReminderTime));
     while (!Application::IsShuttingDown()) {
+        auto UpdateReminderTimeout = ChronoWrapper::TimeFromStringWithLiteral(Application::Settings.getAsString(Settings::Key::Misc_UpdateReminderTime));
         Body = GenerateCall();
         // a hot-change occurs when a setting has changed, to update the backend of that change.
         auto Now = std::chrono::high_resolution_clock::now();
```

We need to do this because UpdateReminderTime is (correctly) read+write, but it's only ever read once. That's broken :)

**Comment by lionkor** [comment]
Also remove line ChronoWrapper.cpp:13 so it doesn't spam ;)

**Comment by lionkor** [comment]
The error message when accessing an  unknown key or category sucks, can we get that to say something like "that's not a setting ya doofus"

**Comment by lionkor** [comment]
Remove ubuntu 20.04 support due to their outdated ahh compiler as well please

**Review by lionkor — APPROVED** [review]
lgtm

**Pull request merged**
